parsers/CY.py

```diff
@@ -12,64 +12,102 @@
 # please try to write PEP8 compliant code (use a linter). One of PEP8's
 # requirement is to limit your line length to 79 characters.
 
-
-def append_datum(data, time_str, prods_str):
-    time_str = time_str[0] + ' ' + time_str[1] + ':' + time_str[2]
-    time_value = arrow.get(time_str).replace(tzinfo='Asia/Nicosia').datetime
-
-    prods_value = {
-        'oil': float(prods_str[3]),
-        'solar': float(prods_str[1]),
-        'wind': float(prods_str[0]),
-        'biomass': 0.0
-    }
-
-    # Because solar is explicitly listed as "Solar PV" (so no thermal with energy storage) and there
-    # is zero sunlight in the middle of the night (https://www.timeanddate.com/sun/cyprus/nicosia),
-    # we use the biomass+solar generation reported at 0:00 to determine the portion of biomass+solar
-    # which constitutes biomass
-    nocturnal_biomass_generation = data[0]['production']['biomass'] if len(data) != 0 else prods_value['solar']
-    prods_value['solar'] -= nocturnal_biomass_generation
-    prods_value['biomass'] += nocturnal_biomass_generation
-    if prods_value['solar'] < 0.0:
-        # if there is (next to) no sunlight and biomass is lower than at midnight
-        prods_value['solar'] = 0.0
-
-    data.append({
-        'zoneKey': 'CY',
-        'production': prods_value,
-        'storage': {},
-        'source': 'tsoc.org.cy',
-        'datetime': time_value
-    })
-
-
-def parse_html(html):
-    html = html.replace('\n', ' ').replace('\r', ' ')
-
-    columns = [m.group(1) for m in re.finditer(
-        '''data\.addColumn\("number", "([^"]*)"\);''', html)]
-    assert columns == ['Συνολική Προβλεπόμενη Ζήτηση', 'Αιολική Παραγωγή',
-        'Εκτίμηση Διεσπαρμένης Παραγωγής (Φωτοβολταϊκά και Βιομάζα)',
-        'Συνολική Ζήτηση', 'Συμβατική Παραγωγή'], 'Source format changed'
-
-    times = (m.group(1, 2, 3) for m in re.finditer(
-        '''var dateStr = "([^"]*)";\s+var hourStr = "(\d+)";\s+var minutesStr = "(\d+)";''', html))
-    prods = (m.group(1, 2, 3, 4) for m in re.finditer(
-        '''\[\s*dateStrFormat,\s*(\d+|null),\s*(\d+|null),\s*(\d+|null),\s*(\d+|null)\s*\]''', html))
-
-    data = []
-    for t, p in zip(times, prods):
-        if any(e == 'null' for e in p):
-            break
-        append_datum(data, t, p)
-
-    return data
+EXPECTED_COLUMNS = ['Συνολική Προβλεπόμενη Ζήτηση', 'Αιολική Παραγωγή',
+    'Εκτίμηση Διεσπαρμένης Παραγωγής (Φωτοβολταϊκά και Βιομάζα)',
+    'Συνολική Ζήτηση', 'Συμβατική Παραγωγή']
+
+COLUMNS_REGEX = re.compile(r'data\.addColumn\("number", "([^"]*)"\);')
+TIMES_REGEX = re.compile(
+    r'var dateStr = "([^"]*)";\s+var hourStr = "(\d+)";\s+var minutesStr = "(\d+)";')
+PRODUCTIONS_REGEX = re.compile(
+    r'\[dateStrFormat,\s*(\d+|null),\s*(\d+|null),\s*(\d+|null),\s*(\d+|null)\]')
+
+class CyprusParser:
+    session = None
+    logger: logging.Logger = None
+
+    def __init__(self, session, logger: logging.Logger):
+        self.session = session
+        self.logger = logger
```

Comment:
There is no real benefit to making this parser a class (the class is holding functions but the object doesn't use them).

Reply:
The idea is that if we were to add logging in more places or make more requests in the future, we wouldn't need to pass these objects as function parameters.

Reply:
That is true but we trade off by having to use `self` everywhere and making the structure of the parser more complicated. However it's not a huge deal, let's see what @FelixDQ thinks.

Reply:
I don't have a strong opinion for or against classes. I think that the implemented solution using a class is neat, so I'm fine with this solution.

I also don't want to be very strict about how a parser should be implemented because I want to be easy for new contributors to create a parser. We obviously have to draw the line somewhere, but I think classes are fine.

Reply:
Alright let's roll with it then.